components/script/dom/htmlformelement.rs

```diff
@@ -599,28 +646,35 @@ impl HTMLFormElement {
         let base = doc.base_url();
         // TODO: Handle browsing contexts (Step 4, 5)
         // Step 6
-        if submit_method_flag == SubmittedFrom::NotFromForm && !submitter.no_validate(self) {
-            if self.interactive_validation().is_err() {
-                // TODO: Implement event handlers on all form control elements
-                self.upcast::<EventTarget>().fire_event(atom!("invalid"));
+        if submit_method_flag == SubmittedFrom::NotFromForm {
+            if self.firing_submission_events.get() {
                 return;
             }
-        }
-        // Step 7
-        // spec calls this "submitterButton" but it doesn't have to be a button,
-        // just not be the form itself
-        let submitter_button = match submitter {
-            FormSubmitter::FormElement(f) => {
-                if f == self {
-                    None
-                } else {
-                    Some(f.upcast::<HTMLElement>())
+            self.firing_submission_events.set(true);
+
+            if !submitter.no_validate(self) {
+                if self.interactive_validation().is_err() {
+                    // TODO: Implement event handlers on all form control elements
+                    self.upcast::<EventTarget>().fire_event(atom!("invalid"));
```

Comment:
I'm seeing the even firing as part of static validation, not interactive validation. See https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constraint-validation:event-invalid

Reply:
I didn't change this part, I merely added indentation. Looking at the code you mentioned, I got an impression that static validation handles only elements inside the form, whereas this line fires an event on the form itself. I might be wrong about this, would you mind double-checking?

Reply:
Ok. You're right about the different event targets, and I actually don't see where in the spec the event is fired on the element as a whole, however if this isn't an actual change, let's leave it for now...

Reply:
Actually I think I've found where the event should be fired on the element as a whole, and this line appears incorrect, so I've filed a follow-up at https://github.com/servo/servo/issues/27133

Reply:
Actually I think I've found where the event should be fired on the element as a whole, and this line appears incorrect, so I've filed a follow-up at https://github.com/servo/servo/issues/27133